Lib/fontTools/ttLib/ttGlyphSet.py

```diff
@@ -285,6 +278,78 @@ def draw(self, pen):
         self.glyphSet.charStrings[self.name].draw(pen, self.glyphSet.blender)
 
 
+class _TTGlyphVARC(_TTGlyph):
+    def _draw(self, pen, isPointPen):
+        """Draw the glyph onto ``pen``. See fontTools.pens.basePen for details
+        how that works.
+        """
+        from fontTools.ttLib.tables.otTables import (
+            VarComponentFlags,
+            NO_VARIATION_INDEX,
+        )
+
+        glyphSet = self.glyphSet
+        varc = glyphSet.varcTable
+        idx = varc.Coverage.glyphs.index(self.name)
+        glyph = varc.VarCompositeGlyphs.VarCompositeGlyph[idx]
+
+        from fontTools.varLib.multiVarStore import MultiVarStoreInstancer
+
+        fvarAxes = glyphSet.font["fvar"].axes
+        instancer = MultiVarStoreInstancer(
+            varc.MultiVarStore, fvarAxes, self.glyphSet.location
+        )
+
+        for comp in glyph.components:
+            location = {}
+            if comp.axisIndicesIndex is not None:
+                axisIndices = varc.AxisIndicesList.Item[comp.axisIndicesIndex]
+                axisValues = Vector(comp.axisValues)
+                if comp.axisValuesVarIndex != NO_VARIATION_INDEX:
+                    axisValues += fi2fl(instancer[comp.axisValuesVarIndex], 14)
+                assert len(axisIndices) == len(axisValues), (
+                    len(axisIndices),
+                    len(axisValues),
+                )
+                location = {
+                    fvarAxes[i].axisTag: v for i, v in zip(axisIndices, axisValues)
+                }
+
+            if comp.transformVarIndex != NO_VARIATION_INDEX:
+                deltas = instancer[comp.transformVarIndex]
+                comp = copy(comp)  # Shallow copy
+                comp.applyTransformDeltas(deltas)
+            transform = comp.transform
+
+            reset = comp.flags & VarComponentFlags.RESET_UNSPECIFIED_AXES
+            with self.glyphSet.glyphSet.pushLocation(location, reset):
+                with self.glyphSet.pushLocation(location, reset):
+                    try:
+                        pen.addVarComponent(
+                            comp.glyphName, transform, self.glyphSet.rawLocation
+                        )
+                    except AttributeError:
+                        t = transform.toTransform()
+                        compGlyphSet = (
+                            self.glyphSet
+                            if comp.glyphName != self.name
+                            else glyphSet.glyphSet
+                        )
```

Comment:
is this the part where you guard against the infinite recursion?

https://github.com/harfbuzz/boring-expansion-spec/blob/main/VARC.md#processing

> if a component refers to the glyphID of the current glyph, instead of recursing (which would result in infinite recursion), the glyph outline for the component is loaded directly instead, as if the glyphID had no VARC entry.

Isn't it possible that a cycle occurs not immediately from current glyph to a direct component, but indirectly through some additional steps in the chain of component references?
E.g. a varcomposite glyph reference some varcomponent, whose base glyph is in turn a varcomposite which has a varcomponent referencing the original varcomposite, etc...

Reply:
This is NOT to prevent a cycle. Cycle detection needs to be implemented separately. I don't think we do it for the regular components either.

This is specifically to allow mixing outlines with components, but allowing a VARC glyphid to refer to its own glyf glyphid.

Reply:
gotcha